flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessor.java

```diff
@@ -0,0 +1,309 @@
+import org.apache.flink.streaming.runtime.tasks.OperatorChain;
+import org.apache.flink.util.ExceptionUtils;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Input processor for {@link MultipleInputStreamOperator}.
+ */
+@Internal
+public final class StreamMultipleInputProcessor implements StreamInputProcessor {
+
+	private final MultipleInputSelectionHandler inputSelectionHandler;
+
+	private final InputProcessor<?>[] inputProcessors;
+
+	private final OperatorChain<?, ?> operatorChain;
+
+	/**
+	 * Stream status for the two inputs. We need to keep track for determining when
+	 * to forward stream status changes downstream.
+	 */
+	private final StreamStatus[] streamStatuses;
+
+	private final Counter numRecordsIn;
+
+	/** Always try to read from the first input. */
+	private int lastReadInputIndex = 1;
+
+	private boolean isPrepared;
+
+	public StreamMultipleInputProcessor(
+			CheckpointedInputGate[] checkpointedInputGates,
+			TypeSerializer<?>[] inputSerializers,
+			IOManager ioManager,
+			StreamStatusMaintainer streamStatusMaintainer,
+			MultipleInputStreamOperator<?> streamOperator,
+			MultipleInputSelectionHandler inputSelectionHandler,
+			WatermarkGauge[] inputWatermarkGauges,
+			OperatorChain<?, ?> operatorChain,
+			Counter numRecordsIn) {
+
+		this.inputSelectionHandler = checkNotNull(inputSelectionHandler);
+
+		List<Input> inputs = streamOperator.getInputs();
+		int operatorsCount = inputs.size();
+
+		this.inputProcessors = new InputProcessor[operatorsCount];
+		this.streamStatuses = new StreamStatus[operatorsCount];
+		this.numRecordsIn = numRecordsIn;
+
+		for (int i = 0; i < operatorsCount; i++) {
+			StreamTaskNetworkOutput dataOutput = new StreamTaskNetworkOutput<>(
+				inputs.get(i),
+				streamStatusMaintainer,
+				inputWatermarkGauges[i],
+				i);
+
+			inputProcessors[i] = new InputProcessor(
+				dataOutput,
+				new StreamTaskNetworkInput<>(
+					checkpointedInputGates[i],
+					inputSerializers[i],
+					ioManager,
+					new StatusWatermarkValve(checkpointedInputGates[0].getNumberOfInputChannels(), dataOutput),
+					i));
+		}
+
+		this.operatorChain = checkNotNull(operatorChain);
+	}
+
+	@Override
+	public CompletableFuture<?> getAvailableFuture() {
+		if (inputSelectionHandler.areAllInputsSelected()) {
+			return isAnyInputAvailable();
+		} else {
+			throw new UnsupportedOperationException();
+		}
+	}
+
+	@Override
+	public InputStatus processInput() throws Exception {
+		int readingInputIndex;
+		if (isPrepared) {
+			readingInputIndex = selectNextReadingInputIndex();
+		} else {
+			// the preparations here are not placed in the constructor because all work in it
+			// must be executed after all operators are opened.
+			readingInputIndex = selectFirstReadingInputIndex();
```

Comment:
Couldn't we introduce some open/init for this case?

Reply:
+/- 0 from my side. Complicating the interface of the class vs a tinny implementation hiccup ?

Reply:
Leave as is. Ideally we would have fully initialized operators before any processor is used.